### ISSUE TITLE
error message when trying to create event; event saves even though error msg says it didn't

### ISSUE THREAD

**Issue by JChipault** [USGS-WiM/whispersservices#475]
Received report on May 7 from a user in Idaho (Tricia Hebdon): "the system threw an error when I was adding our last event and now it is in the system 3 times? Not sure how this happened or what I did."

Neil took a look and found that this event information was actually entered 7 times total; he deleted 6 of the events and kept 201273. He noticed that the time stamps for event creation were spaced by about a minute, which wouldn't have been enough time for Tricia to re-enter the event. We suspect that she clicked "Save", got an error message, clicked "Save" again, got an error message again... repeat until you have 7 copies.

But she didn't relay any more information about what the error message said so we felt we were at a bit of a dead end.

Fast forward to today (May 18)...
Jenny was testing the bulk upload feature and was uploading a CSV with just one event and got this error message:
![image](https://user-images.githubusercontent.com/49287076/118721506-92ed6e80-b7f0-11eb-9a85-f79833f754c4.png)

Tried again and got same error message.

But then when I went to see if the data had actually saved, I did indeed have two copies of the event (201293, 201294).

To try to trouble-shoot more, I entered the event through the Create Event form on the TEST site and it saved fine, no error message (event 171005). But when I tried to use bulk upload on the TEST site, it gave me a JSON error message but did indeed save the event (171006).

It's tempting to say that the issue is with bulk upload but 1) Tricia Hebdon in Idaho was not using bulk upload and 2) I was able to upload a different event (201292) through the bulk upload just fine today.

**Comment by BlakeDraper** [comment]
This is a common problem we have been aware of for a while. It is true - clicking save again would simply re-submit the form data entered, which is exactly how these events got created multiple times. That much I know. What we need to investigate is why the error is coming back from the server when the event submit actually succeeds. We may need to test by entering a new event with the same data as event 201273 and see what in that submission is causing the behavior on the server side.  Any thoughts on this one @aaronstephenson ?

**Comment by aaronstephenson** [comment]
@JChipault  As Blake says, I would need the same JSON that the user submitted in order to test this scenario locally and debug it.

**Comment by BlakeDraper** [comment]
@aaronstephenson since the user would not still have that, I think we can just create it from the event data that exists for the event in question.

**Comment by aaronstephenson** [comment]
Sorry, I was implying that Jenny compile that JSON. I could certainly do it too, I'm just trying to be cognizant of how I spend my remaining hours.

**Comment by JChipault** [comment]
I tried entering the info I see in 201273 as an event in the test site but it saved fine as 171010

**Comment by JChipault** [comment]
Happened again yesterday (5/25). Had new user (Eva in Louisiana) try to create their first event and they got a SyntaxError that read “Unexpected Token T in JSON at position 0”. Turns out that event 201311 had saved anyway but Eva didn't realize that. 

Our staff was trying to help Eva and advised removing the lat/long (since we know we have a problem with that) and Eva said "I tried removing just the coordinates first, and then the coordinates and the location name, but am still getting the same error." (And events 201312-201320 were created as she continued to click "Save Event" since she continued to get the error message). 

Barb at NWHC was able to create a copy of the event yesterday (201321). 

I tried on the test site this morning from a partner manager account (jennypartner1) and it worked fine.

@aaronstephenson **I'd like to delete events 201312-201321 ASAP so that we don't have duplicates in the live system. Do you need those for trouble-shooting or can I delete?**

**Comment by aaronstephenson** [comment]
No, I don't need the duplicates. I'm not working in the prod DB anyway.

**Comment by JChipault** [comment]
@aaronstephenson Got more info from Eva... she said she DID include a service request when she tried to submit data yesterday and hit that token error but I did not see any service request data when I looked closely this morning at event 201311 that she created. I was able to add the service request to 201311 today but, since we've had issues with that service request module and it's associated emails during create event before, I wanted to highlight that it appears that Eva filled out the service request part of the create event form but those data did not save and an email was not issued (as far as I can tell)

**Comment by aaronstephenson** [comment]
I looked at the logs and here's the underlying error:

```
File "/www/htdocs/prod/whispersservices/whispersapi/serializers.py", line 1146, in create
    [new_event_diagnoses.remove(x) for x in new_event_diagnoses if int(x['diagnosis']) in remove_diagnoses]
File "/www/htdocs/prod/whispersservices/whispersapi/serializers.py", line 1146, in <listcomp>
    [new_event_diagnoses.remove(x) for x in new_event_diagnoses if int(x['diagnosis']) in remove_diagnoses]
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

It's a type mismatch, python can't convert a null into an int. Easy fix. I'll look around for other similar situations, too, and fix them if there are any.

**Comment by aaronstephenson** [comment]
Fixed in 40eeb32b and pushed to test server. I found just the one occurrence of that situation. Please try it out when you can.

**Comment by JChipault** [comment]
I wasn't able to reproduce the issue before so not quite sure how to test it. And I don't know code well enough to understand the bug you found. Would that bug explain why the service request data was not saved for 201311 yesterday? 

I did try the bulk upload again (noted previously in this string) and the error message changed to: 
![image](https://user-images.githubusercontent.com/49287076/119723775-75915380-be33-11eb-9846-03ccb9aac2ac.png)
We should be able to upload an event with a diagnosis because it isn't a required field so I'm not sure what's up.

Should we bring Neil in since he might be able to liaison here? This might be over my head but it's a super important bug to fix.

**Comment by aaronstephenson** [comment]
Yes, the error I found in the code explains why the service request was never created.

The error happened at a point in the event creation process immediately prior to the service request creation. So the event itself, plus all child objects (event locations, event collaborators, event organizations, event comments, and event groups) were all created just fine. Then the code encountered an error when processing event diagnoses, which ended all processing, so that no event diagnoses and no service requests were created.

Yes, I know that diagnosis is not required. What you're seeing is an unintended side effect of the fix I just implemented earlier today. I have no experience with the bulk uploaded, since it is entirely client-side code, so I haven't tested that. I'll need to convert your excel data into JSON in order to see what's happening here.

**Comment by JChipault** [comment]
I think I can get you json data but don't know if that'll help with the uploader bug. let me look into that a bit more... be back in touch shortly...

**Comment by JChipault** [comment]
Here's the JSON that I pulled from the "Form data JSON" section of the frontend while logged in as jchipault. The event saved fine though, so not sure this will tell us anything. Fixing the bulk uploader is not nearly as important as just being able to use the Create Event form without error messages so perhaps your fix will do the trick for create event and I create a separate (low priority) ticket for the bulk uploader???

 {
  "event_reference": "SCWDS 20-303",
  "event_type": "1",
  "complete": false,
  "public": true,
  "staff": null,
  "event_status": "1",
  "legal_status": 1,
  "legal_number": "",
  "new_organizations": [
    {
      "org": 914
    }
  ],
  "new_service_request": {
    "request_type": 0,
    "new_comments": []
  },
  "new_event_diagnoses": [],
  "new_comments": [],
  "new_eventgroups": [],
  "new_event_locations": [
    {
      "name": "",
      "start_date": "2020-06-18T05:00:00.000Z",
      "end_date": null,
      "country": "30",
      "administrative_level_one": 10,
      "administrative_level_two": 367,
      "latitude": null,
      "longitude": null,
      "land_ownership": "8",
      "gnis_name": "",
      "gnis_id": "",
      "site_description": "",
      "history": "sadfad",
      "environmental_factors": "",
      "clinical_signs": "",
      "comment": "",
      "event_type": 1,
      "new_location_species": [
        {
          "species": 353,
          "population_count": null,
          "sick_count": null,
          "dead_count": 1,
          "sick_count_estimated": null,
          "dead_count_estimated": null,
          "priority": null,
          "captive": false,
          "age_bias": null,
          "sex_bias": null,
          "new_species_diagnoses": []
        }
      ],
      "new_location_contacts": []
    }
  ],
  "new_read_collaborators": [],
  "new_write_collaborators": []
}

**Comment by aaronstephenson** [comment]
OK, I installed the whispers client on my laptop in order to see what's going on, and got the exact same validation error that you got.

It appears that the bulk uploader _is_ submitting an event diagnosis, with a literal value of "null". This is distinct from an empty value for event diagnosis, which is allowed by the API. To illustrate, this is what the bulk uploader is sending:

![image](https://user-images.githubusercontent.com/8713712/119729381-39152600-be3a-11eb-9f88-daed22dfda0b.png)
in JSON, that event diagnosis would look like this:
```
new_event_diagnoses: [
    "diagnosis": null
]
```

which is not allowed by the API, because to the API this says "I am trying to create a diagnosis whose name is literally 'null'".

The correct way to submit no event diagnosis is like this:
```
new_event_diagnoses: []
```

notice that `new_event_diagnoses` is an empty array (exactly like the example you posted in the immediate previous comment here, which succeeded), not an array with one item whose value is null.

Alternately, you could submit the entire larger JSON body with _no_ `new_event_diagnoses` item at all, and that would also be valid.

@BlakeDraper any thoughts on why `new_event_diagnoses` is being converted from an empty cell into an array with one item by the bulk uploader? My own uninformed guess is that the cell is not empty, and perhaps contains a space or some other 'hidden' character that is being interpreted as a null?

**Comment by JChipault** [comment]
ah, interesting. I just re-checked the .csv file to see if there was a space in the diagnosis cell and it doesn't seem like there was. I hit backspace/delete a lot in that cell. Tried the uploader again and still got that diagnosis error message.

**Comment by BlakeDraper** [comment]
@aaronstephenson @JChipault I found the problem and fixed it. Upcoming in another patch release.

**Comment by JChipault** [comment]
@BlakeDraper Can you let me know when this is ready for testing again?

**Comment by BlakeDraper** [comment]
@JChipault  It is ready now. Sorry - got lost in the notification shuffle since this conversation moved to the backend repository issues.

**Comment by JChipault** [comment]
Still getting this: 

![image](https://user-images.githubusercontent.com/49287076/121196312-fdb12900-c835-11eb-9d90-ab71cd505433.png)

**Comment by BlakeDraper** [comment]
@JChipault  Ugh sorry my bad - I totally lied. I forgot this fix came after those others and actually _isn't_ up on test yet. I apologize for wasting your time there.

**Comment by JChipault** [comment]
@BlakeDraper Bulk upload works now - thanks!

**Comment by aaronstephenson** [comment]
Fix included in v2.1.4 21520fb and deployed to production environment.
